### PR TITLE
sql, server: expose contention info via vtable and StatusServer

### LINKED ISSUE #57114: sql: implement contention registry and expose through StatusServer and virtual table
One of SQL Execution's 21.1 goals is to surface global KV contention information. To do this, each gateway node will have to keep track of contention events for its queries.

This map view is still pending design, but we can start work on the backend given the general idea. This “global” view should also be queryable via a virtual table. This kind of API is reminiscent of other features like `SHOW QUERIES`, where each server has a local view of its queries and there is a cluster-level API through the DB console as well as SQL shell to query the global state.

The proposal in this case is to keep an in-memory contention registry keyed by table ID/indexID pair that is updated by the DistSQL receiver when receiving contention events. When a `StatusServer` receives a request for a global contention view, it will broadcast a contention request to each node for its local contention view, and merge it with its own local view.

Diving a bit deeper, what should this contention registry look like? It depends on the questions we want to answer. At a high level, we want to be able to answer which tables are experiencing contention and allow the user to dive deeper into a table to understand what key/row range is experiencing contention and which transactions are responsible for this contention. Therefore, the proposal is a top-level map keyed by a table ID/index ID with a value struct that looks something like:
```
type tableContentionInfo struct {
    contentionEvents uint64
    cumulativeContentionTime time.Duration
    orderedKeyContentionInfo []keyContentionInfo
}

type keyContentionInfo struct {
    key roachpb.Key
    contendingTransactions []roachpb.Transaction
}
```

This map cannot grow unboundedly, so an eviction policy needs to be put into place. As a simple starter, we can also keep track of the last contention event for a given table and use an LRU policy with a maximum size for the map. In a given map entry, we could similarly use an LRU policy for keys.

Merging these structs should be relatively straightforward.

### PR DESCRIPTION
**Pull request by yuzefovich**
This commit adds two new endpoints to the status server
`contention_events` and `local_contention_events` (as well as
corresponding methods) for using internally that expose the contention
information that comes from `contention.Registry`. These endpoints
require the user to either be an admin or have `VIEWACTIVITY` grant.

Additionally, this commit introduces new
`crdb_internal.{cluster,node}_contention_events` virtual tables that
provide access to the contention info via SQL.

Fixes: #57114.

Release justification: low risk update to new functionality.

Release note (sql change): New virtual tables
`crdb_internal.cluster_contention_events` and
`crdb_internal.node_contention_events` are introduced which expose
cluster-wide and gateway-only, respectively, contention information. In
order to access it the user needs to be either an admin or have
`VIEWACTIVITY` grant.

### CODE REVIEW

**Comment by yuzefovich** [comment]
First two commits are in separate PR, but otherwise I think this is RFAL.

**Review by asubiotto — CHANGES_REQUESTED** [review]
This is so exciting!! @tbg @awoods187 

> Reviewed 6 of 6 files at r1, 7 of 7 files at r2, 29 of 29 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-:-MUEGG0I0ApQ_azoZKSQ:bodeqho)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/server/status.go, line 307 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-MUEHktO2L9zkS3p2C3k:-MUEHktO2L9zkS3p2C3l:b-92nm3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/1244fa841beddced68dbb77010e441589c9ee92c/pkg/server/status.go#L307)):*
> ```Go
> 	ctx = b.AnnotateCtx(ctx)
> 
> 	sessionUser, isAdmin, err := b.privilegeChecker.getUserAndRole(ctx)
> ```

not your code, but I'm surprised that this `VIEWACTIVITY` check is not extracted since I think a lot of these methods probably care about it and it's pretty repetitive. Do you mind extracting this privilege check (and the error return)?.
___
*[pkg/server/status.go, line 1852 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-MUEGL5H1pdZkWjUlG5Z:-MUEGL5H1pdZkWjUlG5_:bt5svp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1244fa841beddced68dbb77010e441589c9ee92c/pkg/server/status.go#L1852)):*
> ```Go
> }
> 
> // ListLocalContentionEvents returns a list of contention events on this node.
> ```

nit: add ordering information since this is also included in the docs.
___
*[pkg/server/status.go, line 2062 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-MUEI2Gx1JcZVLH240sk:-MUEI2Gx1JcZVLH240sl:bn8raq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1244fa841beddced68dbb77010e441589c9ee92c/pkg/server/status.go#L2062)):*
> ```Go
> }
> 
> func (s *statusServer) listContentionEventsHelper(
> ```

Is this used anywhere other than `ListContentionEvents`? Since there's not much in there, I think it's fine to put the code in `ListContentionEvents` and remove this separate method.
___
*[pkg/server/status.go, line 2079 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-MUEIFzE8p3bJq_a0nNO:-MUEIFzE8p3bJq_a0nNP:b-p4ln7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1244fa841beddced68dbb77010e441589c9ee92c/pkg/server/status.go#L2079)):*
> ```Go
> 			return resp.Events, nil
> 		}
> 		return nil, err
> ```

nit: put error case inside if.
___
*[pkg/server/status.go, line 2193 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-MUEIWqb7VjbIkCTE_4e:-MUEIWqb7VjbIkCTE_4f:b-4h423a) ([raw file](https://github.com/cockroachdb/cockroach/blob/1244fa841beddced68dbb77010e441589c9ee92c/pkg/server/status.go#L2193)):*
> ```Go
> 	ctx = s.AnnotateCtx(ctx)
> 
> 	if _, _, err := s.privilegeChecker.getUserAndRole(ctx); err != nil {
> ```

Does this call do anything?
___
*[pkg/sql/crdb_internal.go, line 1690 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-MUEJUdHBvQ6XRZqVB3V:-MUEJUdHBvQ6XRZqVB3W:b-8n23bn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1244fa841beddced68dbb77010e441589c9ee92c/pkg/sql/crdb_internal.go#L1690)):*
> ```Go
> 
> const contentionEventsSchemaPattern = `
> CREATE TABLE crdb_internal.%s (
> ```

We should probably inform the caller of the eviction policy in the virtual table comment field.
___
*[pkg/sql/logictest/testdata/logic_test/crdb_internal, line 215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-MUEJod44laUwfmaBOPm:-MUEJod44laUwfmaBOPn:b-wytga9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1244fa841beddced68dbb77010e441589c9ee92c/pkg/sql/logictest/testdata/logic_test/crdb_internal#L215)):*
> ```
> 
> query IIITTTI colnames
> SELECT * FROM crdb_internal.node_contention_events WHERE table_id < 0
> ```

I think there are also some logic tests that @tbg added to produce contention events. Would probably be good to check that those contention events make it into the virtual table.

**Comment by tbg** [comment]
Wohoo! Can we have a sample printout of that vtable pretty please? :-)

**Comment by knz** [comment]
As discussed elsewhere, the implementation approach taken here is amplifying the problem identified in #60584  and so we should try to find a different approach instead.

**Review by yuzefovich — COMMENTED** [review]
@tbg, here is an example (I ran a single node cluster and followed the queries of `contention_event` logic test file):
```
root@:26257/defaultdb> select * from crdb_internal.cluster_contention_events;
  table_id | index_id | num_contention_events | cumulative_contention_time |                      key                      |                txn_id                | count
-----------+----------+-----------------------+----------------------------+-----------------------------------------------+--------------------------------------+--------
        64 |        1 |                     1 | 00:00:00.083261            | \310\211\375\010\323R\200\013\212\000\001\210 | 7e9f5456-ed01-4525-97d1-a72dfd06a339 |     1
(1 row)

Time: 32ms total (execution 32ms / network 0ms)

root@:26257/defaultdb> select * from crdb_internal.node_contention_events;
  table_id | index_id | num_contention_events | cumulative_contention_time |                      key                      |                txn_id                | count
-----------+----------+-----------------------+----------------------------+-----------------------------------------------+--------------------------------------+--------
        64 |        1 |                     1 | 00:00:00.083261            | \310\211\375\010\323R\200\013\212\000\001\210 | 7e9f5456-ed01-4525-97d1-a72dfd06a339 |     1
(1 row)

Time: 0ms total (execution 0ms / network 0ms)
```

> As discussed elsewhere, the implementation approach taken here is amplifying the problem identified in #60584 and so we should try to find a different approach instead.

Ack that the current approach is broken, but it will have to be fixed at the same time as existing violations are.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-:-MUHWy5h25dnsU-OiRBF:b-4c8cax)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @itsbilal, and @tbg)

___
*[pkg/server/status.go, line 307 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-MUEHktO2L9zkS3p2C3k:-MUHKxgD3bBLqmaUtqd5:bshazvx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1244fa841beddced68dbb77010e441589c9ee92c/pkg/server/status.go#L307)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

not your code, but I'm surprised that this `VIEWACTIVITY` check is not extracted since I think a lot of these methods probably care about it and it's pretty repetitive. Do you mind extracting this privilege check (and the error return)?.
</blockquote></details>

Actually, `VIEWACTIVITY` is used only for `SESSIONS`, and there we have special handling that allows any user to see its own sessions (even if that other is not admin nor has VIEWACTIVITY permissions), so I won't be touching any of the old code.
___
*[pkg/server/status.go, line 1852 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-MUEGL5H1pdZkWjUlG5Z:-MUHHrqSCF1MtyE9IN2p:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/1244fa841beddced68dbb77010e441589c9ee92c/pkg/server/status.go#L1852)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: add ordering information since this is also included in the docs.
</blockquote></details>

Good point, done.
___
*[pkg/server/status.go, line 2062 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-MUEI2Gx1JcZVLH240sk:-MUHIbd94aNG4_W-cHhQ:b-jxan70) ([raw file](https://github.com/cockroachdb/cockroach/blob/1244fa841beddced68dbb77010e441589c9ee92c/pkg/server/status.go#L2062)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Is this used anywhere other than `ListContentionEvents`? Since there's not much in there, I think it's fine to put the code in `ListContentionEvents` and remove this separate method.
</blockquote></details>

Refactored. I originally implemented api V2 (which was using the helper) but decided to get rid of it.
___
*[pkg/server/status.go, line 2193 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-MUEIWqb7VjbIkCTE_4e:-MUHJ7_kBo04w7LOMLyb:b3ett55) ([raw file](https://github.com/cockroachdb/cockroach/blob/1244fa841beddced68dbb77010e441589c9ee92c/pkg/server/status.go#L2193)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Does this call do anything?
</blockquote></details>

Hm, good question. I copy-pasted this from `ListSessions`, but it appears that it doesn't do much other than checking whether a user is specified in the context.

I'm guessing maybe the reasoning is that we will issue `ListLocalContentionEvents` RPC for each node, and each node locally performs the permissions check, so if the user doesn't have permissions, we should be getting errors from all of the nodes. However, I think we could do that check upfront and fail quickly if not enough permissions. cc @itsbilal to see whether my reasoning is correct.
___
*[pkg/sql/crdb_internal.go, line 1690 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-MUEJUdHBvQ6XRZqVB3V:-MUHQH7_Ebi4sOiji3b0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1244fa841beddced68dbb77010e441589c9ee92c/pkg/sql/crdb_internal.go#L1690)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

We should probably inform the caller of the eviction policy in the virtual table comment field.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/crdb_internal, line 215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-MUEJod44laUwfmaBOPm:-MUHWsQd5lNHwuig9cWw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1244fa841beddced68dbb77010e441589c9ee92c/pkg/sql/logictest/testdata/logic_test/crdb_internal#L215)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think there are also some logic tests that @tbg added to produce contention events. Would probably be good to check that those contention events make it into the virtual table.
</blockquote></details>

Done.

**Review by asubiotto — APPROVED** [review]
:lgtm: 

> As discussed elsewhere, the implementation approach taken here is amplifying the problem identified in #60584 and so we should try to find a different approach instead.

Ack on my end as well. I agree there is a problem, but this change does not break anything new, so it only slightly increases the existing technical debt by introducing one more API call to be moved. Since we're so close to release, I think the extremely high benefit of this change (allowing users to observe contention) justifies merging this.

> Reviewed 11 of 11 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-:-MUJF0uf69ssY23yA0LG:b37fkku)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @tbg, and @yuzefovich)

___
*[pkg/server/status.go, line 1868 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-MUJFcvS7uf32J6qHhOW:-MUJFcvS7uf32J6qHhOX:bbn7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/305baefd1441a884d8060ea3ea94bf5a884b8ba8/pkg/server/status.go#L1868)):*
> ```Go
> // On the middle level, all SingleKeyContention objects are ordered by their
> // keys lexicographically.
> // On the lowest level, all SingleTxnContention objects are ordered by the
> ```

nit: maybe it's just me, but I feel like this is a little hard to understand. Maybe reword to something like`ordered by the number of times that transaction was observed to contend with other transactions`. Up to you.

**Review by yuzefovich — COMMENTED** [review]
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60713#-:-MULjZx75Fs8Fiu3nvyt:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @tbg)

**Comment by yuzefovich** [comment]
Unrelated flake.

bors r+

**Comment by yuzefovich** [comment]
Flaked on #61091. Let's try one more time.

bors r+
